### ISSUE TITLE
transform_algo = affine generates poor model.transform()

### ISSUE THREAD

**Issue by hamid-bolouri** [AmirAlavi/scipr#8]
* scipr version: 0.2.1
* Python version: 3.9.7
* Operating System: Mac OS Monterey

### Description

Describe what you were trying to get done.
I 'd like to adapt and use scipr with flow and mass cytomtery data sets.

Tell us what happened, what went wrong, and what you expected to happen.
Using simulated data, MNN matching, and an "affine" transform, the transformed object ends up being a single point repeated N times.  If I use "rigid", I get points on a curve.

### What I Did
As input, I simulated a toy 2D data set containing 500 points in  4 clusters (data set "A" in the code below). Then I affine-transformed "A" and called that "B". 

Paste the command(s) you ran and the output.

import scipr
from scipr.matching import Closest, Hungarian, Greedy, MNN
from scipr.transform import Affine, Rigid

closest = Closest()
MNN = MNN()
rigid = Rigid()
affine = Affine()

model = scipr.SCIPR(match_algo=MNN,
                    transform_algo=affine,
                    input_normalization='l2',
					n_iter = 100)
model.fit(A, B)
aligned = model.transform(A)

I get back a single coordinate:
aligned[:5,]
array([[0.71251583, 0.68878496],
       [0.71251583, 0.68878496],
       [0.71251583, 0.68878496],
       [0.71251583, 0.68878496],
       [0.71251583, 0.68878496]])

Part of the issue is the matching. If I replace the "match_alg"o with one that just passes a predefined set of matching points, then I get back a transformed object that looks reasonably like the reference, but ONLY if the "transform_algo" is "rigid". For "affine" I get back a straight-line. Any ideas why? 

THANK YOU.

Hamid

If there was a crash, please include the traceback here.
```

**Comment by AmirAlavi** [comment]
Hi Hamid,

Thanks for checking out scipr and giving it a shot on your problem! Is this data that you could share? The performance of scipr (and most alignment methods) is heavily dependent on the particular dataset you are working on.

It would help to take a look at your specific case to investigate whether there is a software bug, or rather, a case where applying scipr is not a good option (e.g. the assumptions in scipr are too strong or something like that).

Without seeing the specific data, I can only give some general comments:

- Have you investigated different input normalizations? i.e. The `input_normalization` parameter of the `SCIPR` constructor. There are a few options built in
- How does the algorithm behave over iterations on your data? For example, does it immediately converge after a single iteration, with stable matchings, or does the set of matched pairs fluctuate and change often? 
  - You can get some info from turning on [logging ](https://scipr.readthedocs.io/en/latest/intro.html#logging)
  - Or by turning on [tensorboard](https://scipr.readthedocs.io/en/latest/intro.html#logging) (set `tensorboard=True` in the call to `SCIPR.fit`)

Happy to take a closer look, or brainstorm with you on modifications to make to the library to better suit this type of data.

Cheers,
Amir

**Comment by hamid-bolouri** [comment]
Thanks very much for the explanations and suggestions Amir.  

I think you put your finger on it right away: The issue seems to be the number of iterations. Below is a complete reproducible example using small toy datasets that mimic my intended use case. 

(I generated the attached input files in R by first making dataset B and then affine transforming B to create A. The indices mark my selection of matching anchor points, and the labels mark the 4 clusters in the data.)

In this example, "n_iter = 1" and "n_iter = 2" produce plausible alignments, but higher values of "n_iter" increasingly push all points onto a line.  Maybe something to do with the stopping criteria?

Thanks again!

Hamid

PS. I'd be happy to share examples of actual cytomtery data, but the files are large (30-50 features X 100K-500K cells) and a distraction, since I want to align only specifically-selected anchor-points (optimal number to be determined, but likely O(100)).

```
import pandas as pd
import numpy as np

DIR = "/Users/hbolouri/...etc...I/cytoFusion/"

A = pd.read_csv(DIR  +"unixXY.txt", header=None)
A = A.to_numpy()
B = pd.read_csv(DIR + "unixXY2.txt", header=None)
B = B.to_numpy()

lblsA = pd.read_csv(DIR + "unixLblsXY.txt", header=None)
lblsA = lblsA.to_numpy()
lblsA = lblsA[:, 0]
lblsB = pd.read_csv(DIR + "unixLblsXY2.txt", header=None)
lblsB = lblsB.to_numpy()
lblsB = lblsB[:, 0]

indxA = pd.read_csv(DIR + "unixIndxXY.txt", header=None)
indxA = indxA.to_numpy()
indxA = indxA[:,0] - 1 # switch to python zero start coord
indxB = pd.read_csv(DIR + "unixIndxXY2.txt", header=None)
indxB = indxB.to_numpy()
indxB = indxB[:,0] - 1 # switch to python zero start coord


import scipr
from scipr.matching import Closest, Hungarian, Greedy, MNN
from scipr.transform import Affine, Rigid

closest = Closest()
MNN = MNN()
rigid = Rigid()
affine = Affine()

class HB_match(scipr.matching.Match):
    """Simply returns indices and
    distances of pre-determined 'anchor' cells.
    """
    def __init__(self, A, B, kdB):
        self.tst = "Hi"
    
    def match(self, A, B, kdB):
        distances = spatial.distance.cdist(B[indxB,:], A[indxA,:])
        distances = distances.diagonal() 
        
        return np.array(indxA), np.array(indxB), np.array(distances)


from scipy import spatial
kdB = spatial.cKDTree(B)

myMatch = HB_match(A, B, kdB)

model = scipr.SCIPR(match_algo = myMatch,
                    transform_algo = affine,
                    input_normalization = 'std',
                    n_iter = 100)

model.fit(A, B)
aligned = model.transform(A)


import matplotlib.pyplot as plt
colormap = np.array(['y', 'r', 'g', 'b', 'c'])
plt.style.use('seaborn-whitegrid')

fig = plt.scatter(aligned[:, 0], aligned[:, 1],
                  s=30, c=colormap[lblsA],  alpha=0.25)
plt.show()
```


[unixIndxXY2.txt](https://github.com/AmirAlavi/scipr/files/7583531/unixIndxXY2.txt)
[unixIndxXY.txt](https://github.com/AmirAlavi/scipr/files/7583532/unixIndxXY.txt)
[unixLblsXY2.txt](https://github.com/AmirAlavi/scipr/files/7583533/unixLblsXY2.txt)
[unixLblsXY.txt](https://github.com/AmirAlavi/scipr/files/7583534/unixLblsXY.txt)
[unixXY2.txt](https://github.com/AmirAlavi/scipr/files/7583535/unixXY2.txt)
[unixXY.txt](https://github.com/AmirAlavi/scipr/files/7583536/unixXY.txt)

**Comment by AmirAlavi** [comment]
Hi Hamid, thanks very much for sharing the sample data and the script. 

First, my apologies that you aren't having the results you'd hope for with scipr, and see confusing results. Below I discuss the reasons, and have some suggestions.

### TLDR:

- If you want to use scipr (you want to learn a global transformation function), use Rigid, it is more conservative and easier to trust
- Else, try point-wise methods, namely MNN_correct (but these don't give you a global transformation function)

You can see the alignments on the toy data below from these, and I think Rigid is the clear winner (but this is a toy example), and MNN_correct (a completely different method from scipr) is not bad and could be better on real data.  Affine isn't a good option.

### Details:
In two dimensions, I'd actually expect scipr's affine function to return points on a line. This is expected because it's fitting a linear function on the data to minimize the pairwise distances. It's an overly strong assumption that a single, global affine transformation function can describe the adjustment necessary to align the data.

In scRNA-seq data,  it's very high dimensional (thousands of genes), and we usually make the assumption that for datasets that consist of similar cell types, the data does lie on some lower-dimensional manifold. But it's still probably too strong to say that that manifold is a plane. To be clear, scipr's Affine transform will put data on some plane in high dimensions. In our paper, we saw that this very simple assumption can align some datasets.

However, we also saw that Rigid, surprisingly, actually can align datasets with decent performance (but didn't try it on enough experiments to make conclusions for the paper. There's one example performance plot in the supplement showing Rigid).

**<ins>Honestly, after working on this problem more since the publication,  if using scipr myself, I'd trust Rigid the most.</ins>** 

A big challenge with methods that target the same niche as scipr is that it's hard to do these two things simultaneously:

1. Align the data in the original dimensions
2. Learn a single, global transformation function that can do it

**<ins>I'd also suggest exploring some other approaches too.</ins>**

Instead of trying to learn a single transformation function to align the data, there are other methods that operate point-wise. As in, don't use the same function to transform every point. Learn a function per point. But each function can be simple. **<ins>This is obviously more powerful than scipr</ins>**, but doesn't give you a fitted function you can apply on new data.

For example, have you tried MNN Correct? [paper](https://www.nature.com/articles/nbt.4091)

I've run it on your toy data example, and you can see the plots below with the results. It's maybe not as clean as Rigid, but this is toy data, so worth shot on your real data.

Plots (your A and B, with the pairs you assigned as dotted lines connecting 
them):

![unaligned](https://user-images.githubusercontent.com/2713034/143376760-e35901a7-2b37-41d7-b162-d41dc4fd87a9.png)
![scipr_mnn_affine_aligned](https://user-images.githubusercontent.com/2713034/143376775-73ad0bd4-2e30-4186-b9d7-746994a77e6b.png)
![scipr_mnn_rigid_aligned](https://user-images.githubusercontent.com/2713034/143376794-5d721380-f86a-4818-8e55-4af00d7571a3.png)
![MNN_correct_aligned](https://user-images.githubusercontent.com/2713034/143376804-cacef2eb-0cb3-4019-baac-141ff76fd062.png)



source code to produce the above plots:
```python
import pandas as pd
import numpy as np
from scipy import spatial

import scipr
from scipr.matching import MNN
from scipr.transform import Affine, Rigid

import scanpy as sc # For using MNN_correct
import anndata

import matplotlib.pyplot as plt

def plot(A, B, indxA, indxB, lblsA, lblsB, title, filename):
    fig, axs = plt.subplots(1, 2, figsize=(12, 5))
    axs[0].scatter(A[:, 0], A[:, 1],
                    s=30, alpha=0.25, label='A')
    axs[0].scatter(B[:, 0], B[:, 1],
                    s=30, alpha=0.25, label='B')
    axs[0].legend()
    
    axs[1].scatter(A[:, 0], A[:, 1],
                    s=30, alpha=0.25, c=colormap[lblsA])
    axs[1].scatter(B[:, 0], B[:, 1],
                    s=30, alpha=0.25, c=colormap[lblsB])
    for anchorA, anchorB in zip(indxA, indxB):
        xcords = [A[anchorA, 0], B[anchorB, 0]]
        ycords = [A[anchorA, 1], B[anchorB, 1]]
        axs[0].plot(xcords, ycords, color='k', linestyle='dashed', alpha=0.1)
        axs[1].plot(xcords, ycords, color='k', linestyle='dashed', alpha=0.1)
    plt.suptitle(title)
    plt.savefig(filename)

if __name__ == '__main__':

    DIR = "data/"

    A = pd.read_csv(DIR  +"unixXY.txt", header=None)
    A = A.to_numpy()
    B = pd.read_csv(DIR + "unixXY2.txt", header=None)
    B = B.to_numpy()

    lblsA = pd.read_csv(DIR + "unixLblsXY.txt", header=None)
    lblsA = lblsA.to_numpy()
    lblsA = lblsA[:, 0]
    lblsB = pd.read_csv(DIR + "unixLblsXY2.txt", header=None)
    lblsB = lblsB.to_numpy()
    lblsB = lblsB[:, 0]

    indxA = pd.read_csv(DIR + "unixIndxXY.txt", header=None)
    indxA = indxA.to_numpy()
    indxA = indxA[:,0] - 1 # switch to python zero start coord
    indxB = pd.read_csv(DIR + "unixIndxXY2.txt", header=None)
    indxB = indxB.to_numpy()
    indxB = indxB[:,0] - 1 # switch to python zero start coord

    colormap = np.array(['y', 'r', 'g', 'b', 'c'])

    # Visualize the unaligend data, with HB's assigned pairs
    plot(A, B, indxA, indxB, lblsA, lblsB, "Unaligned data", "unaligned.png")

    MNN = MNN()
    affine = Affine()
    rigid = Rigid()

    class HB_match(scipr.matching.Match):
        """Simply returns indices and
        distances of pre-determined 'anchor' cells.
        """
        def __init__(self, A, B, kdB):
            self.tst = "Hi"
        
        def match(self, A, B, kdB):
            distances = spatial.distance.cdist(B[indxB,:], A[indxA,:])
            distances = distances.diagonal() 
            
            return np.array(indxA), np.array(indxB), np.array(distances)

    kdB = spatial.cKDTree(B)

    myMatch = HB_match(A, B, kdB)

    # Try affine
    model = scipr.SCIPR(match_algo = myMatch,
                        transform_algo = affine,
                        input_normalization = 'std',
                        n_iter = 100)

    model.fit(A, B)
    aligned_affine = model.transform(A)

    # Try rigid
    model = scipr.SCIPR(match_algo = myMatch,
                        transform_algo = rigid,
                        input_normalization = 'std',
                        n_iter = 100)

    model.fit(A, B)
    aligned_rigid = model.transform(A)

    B_normalized = model._apply_input_normalization(B)

    plot(aligned_affine, B_normalized, indxA, indxB, lblsA, lblsB, "Aligned (scipr, affine)", "scipr_mnn_affine_aligned.png")
    plot(aligned_rigid, B_normalized, indxA, indxB, lblsA, lblsB, "Aligned (scipr, rigid)", "scipr_mnn_rigid_aligned.png")

    # Now try MNN_correct
    A_adata = anndata.AnnData(A)
    B_adata = anndata.AnnData(B)

    mnn_args = {
        'cos_norm_in': True,
        'cos_norm_out': False, # The default is to set this to True, but places things on line-like manifolds because of the cosine norm
    }
    corrected, _, _ = sc.external.pp.mnn_correct(B_adata, A_adata, batch_categories=['B', 'A'], **mnn_args)

    
    A_adata_corrected = corrected[corrected.obs['batch'] == 'A'].X
    B_adata_corrected = corrected[corrected.obs['batch'] == 'B'].X

    plot(A_adata_corrected, B_adata_corrected, indxA, indxB, lblsA, lblsB, "Aligned (MNN_correct)", "MNN_correct_aligned.png")

```
